Add topic flag to create robot

ValerioMagnago opened the pull request:
Fixes #127 

## Proposed Changes

  - add new --topic parameter that enables to specify the name of the topic where to read the robot description
  - add a demo launch file that loads the robot description on the `\robot_description` topic by leveraging the `robot_state_publisher ` and then use the newly defined flag to spawn the model in gazebo and additionally the launch starts rviz2 loading the robot description
  -  check `rclcpp::ok()` while server time to download resources. In case of failure the program will return -1

AndrejOrsula committed:
Add default value for plugin path in launch script (#125)

Signed-off-by: Andrej Orsula <orsula.andrej@gmail.com>
Signed-off-by: Valerio Magnago <valerio.magnago@hotmail.it>

ValerioMagnago committed:
chore: Format the code using clang-format

Signed-off-by: Valerio Magnago <valerio.magnago@hotmail.it>

ValerioMagnago committed:
fix: Check rclcpp's status inside loop

now the program should terminate with error when sending termination
signal to the node while waiting ignition gazebo server to download the
resource

Signed-off-by: Valerio Magnago <valerio.magnago@hotmail.it>

ValerioMagnago committed:
feat: Create from topic

Now it is possible to run ros_ign_gazebo create specifying a topic as
source of the robot description

Signed-off-by: Valerio Magnago <valerio.magnago@hotmail.it>

ValerioMagnago committed:
feat: Add demos create from topic

Add a launch file starting a ignition gazebo world and spawn a sphere in it.
Additionally a rviz2 interface is loaded to show that also Rviz can load
the robot description

Signed-off-by: Valerio Magnago <valerio.magnago@hotmail.it>

ValerioMagnago committed:
chore: Remove not useful prints

Signed-off-by: Valerio Magnago <valerio.magnago@hotmail.it>

ValerioMagnago committed:
chore: Includes reordered

Signed-off-by: Valerio Magnago <valerio.magnago@hotmail.it>

chapulina commented on ros_ign_gazebo_demos/launch/robot_description_publisher.launch.py:
```diff
@@ -0,0 +1,90 @@
+    # Robot state publisher
+    params = {'use_sim_time': True, 'robot_description': robot_desc}
+    robot_state_publisher =  Node(
+            package='robot_state_publisher',
```
Let's add `robot_state_publisher` as an `exec_depend` of `ros_ign_gazebo_demos` so it's installed with it

ValerioMagnago replied:
Sure! I forgot it

chapulina commented on ros_ign_gazebo_demos/launch/robot_description_publisher.launch.py:
```diff
@@ -0,0 +1,90 @@
+# Copyright 2019 Open Source Robotics Foundation, Inc.
```
Thanks for the demo. Mind adding an item for it to `ros_ign_gazebo_demos/README.md`?

ValerioMagnago replied:
Sure I will add it.
Is the name of the launch file good? Maybe you have something better in mind?

chapulina replied:
Yeah that sounds good to me, thanks!

chapulina commented on ros_ign_gazebo/src/create.cpp:
```diff
@@ -98,27 +98,49 @@ int main(int _argc, char ** _argv)
+    } while (rclcpp::ok() && future_ret != rclcpp::FutureReturnCode::SUCCESS);
+
+    if (future_ret == rclcpp::FutureReturnCode::SUCCESS) {
+      RCLCPP_INFO(ros2_node->get_logger(), xml_future.get());
```
This could be huge depending on the model being spawned. Consider omitting this info here.

ValerioMagnago replied:
Yes you are right!

chapulina commented on ros_ign_gazebo_demos/launch/robot_description_publisher.launch.py:
```diff
@@ -0,0 +1,90 @@
+        '<inertia ixx="1" ixy="0.0" ixz="0.0" iyy="1" iyz="0.0" izz="1"/>'\
+        '</inertial>'\
+        '</link>'\
+        '</robot>" '
```
This is spawning on RViz, but failing on Gazebo. Removing this stray quote fixes it for me:

```suggestion
        '</robot>'
```

chapulina commented on ros_ign_gazebo_demos/launch/robot_description_publisher.launch.py:
```diff
@@ -0,0 +1,90 @@
+    pkg_ros_ign_gazebo = get_package_share_directory('ros_ign_gazebo')
+    gazebo = IncludeLaunchDescription(
+        PythonLaunchDescriptionSource(
+            os.path.join(pkg_ros_ign_gazebo, 'launch', 'ign_gazebo.launch.py'))
```
This is opening a blank world without any entities. Consider using `empty.sdf`, which has a sun and a ground plane:

```suggestion
            os.path.join(pkg_ros_ign_gazebo, 'launch', 'ign_gazebo.launch.py')),
        launch_arguments={'ign_args': '-r empty.sdf'}.items(),
```

chapulina reviewed (commented):
Thanks for the PR, it works for me. I just have some minor comments below.

CC @clalancette in case you want to take a look too.

ValerioMagnago committed:
fix: Add pkg dependency on robot_state_publisher

The newly created demo introduce a dependency on the robot_state_publisher package

Signed-off-by: Valerio Magnago <valerio.magnago@hotmail.it>

ValerioMagnago committed:
chore: Remove excessive debug message

Now the loaded urdf is no more logged to the console

Signed-off-by: Valerio Magnago <valerio.magnago@hotmail.it>

ValerioMagnago committed:
feat: Better demo rendering

Add a base world with light and ground plane as base environment for the demo

Co-authored-by: Louise Poubel <louise@openrobotics.org>
Signed-off-by: Valerio Magnago <valerio.magnago@hotmail.it>

ValerioMagnago committed:
fix: Fix copy paste error

Co-authored-by: Louise Poubel <louise@openrobotics.org>
Signed-off-by: Valerio Magnago <valerio.magnago@hotmail.it>

ValerioMagnago committed:
docs: Add demo documentation

Add short docs of the demo on the README file and remove unuseful
element from rviz configs.

Signed-off-by: Valerio Magnago <valerio.magnago@hotmail.it>

ValerioMagnago commented:
@chapulina is there anything else that I can do here to help?

chapulina reviewed (approved):
Thanks for the new feature and the demo, works great for me :+1:
